File: neetcode/old/150stringEncodeDecode.py
from typing import List


def encode(strs: List[str]) -> str:
    res = ""
    for s in strs:
        res += str(len(s)) + "#" + s
    return res


def decode(str: str) -> List[str]:
    res = []
    i = 0

    while i < len(str):
        j = str.index("#", i)
        length = int(str[i:j])
        next_str = str[j + 1 : j + 1 + length]
        res.append(next_str)
        i = j + 1 + length
    return res


input_strs = ["I", "code", "for", "fun"]
encoded = encode(input_strs)
print(encoded)
decoded = decode(encoded)
print(decoded)

File: neetcode/old/150stringEncodeDecode.ts
/**
 * @param {string[]} strs
 * @returns {string}
 */

const encode = (strs: string[]): string => {
  let res = '';

  strs.map((str) => (res += str.length + '#' + str));
  console.log(res);

  return res;
};

/**
 * @param {string} str
 * @returns {string[]}
 */

const decode = (str: string): string[] => {
  let res: string[] = [],
    i = 0;

  while (i < str.length) {
    const j = str.indexOf('#', i);
    const len = parseInt(str.slice(i, j), 10);
    console.log(len);

    const nextStr = str.slice(j + 1, j + 1 + len);
    res.push(nextStr);

    i = j + 1 + len;

    console.log(res);
  }

  return res;
};

const input = ['neet', 'code', 'love', 'you'];
const encoded = encode(input);
console.log(encoded);
const decoded = decode(encoded);
console.log(decoded);
